### PR TITLE
Test: K8st update how cilium is installed.

### PR DESCRIPTION
**Pull request by eloycoto**
Before we used a couple of manifest, with the entry of new features like
`--auto-routing` and istio we need a couple more. This commit added a
new third party tool (kubecfg[0]) that allow us to use template manifest
using jsonnet[1] format.

This allow us to reuse the templates, and change how things are deployed
in a easier way.

[0] https://github.com/ksonnet/kubecfg
[1] https://jsonnet.org/

On the other hand added a new ginkgo flag to enable a new manifest to
run the test.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4219)

### CODE REVIEW

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
Hi!

Before you ask, let me explain why I use `kubecfg` instead of any other
solutions, I  wanted to clarify why I choose this one:

- Using Golang templates:  This was my first thought but to find the
differences in the Json manifest was super difficult. If we use yaml, the issue
with the spaces, and for loops is a bit hard. We are speaking about 500 lines
manifest. On the other hand, if you want to try manually is super difficult,
and you always need to run the test to have cilium installed.

- Using sed: This is another option, but I found super difficult to change the
containers args, and using sed, you never know what changes and it's not
explicit. I think that all the test should be explicit.

- Using Kubecfg:  This is a tool designed for this kind of operations. The
cilium manifest is on a jsonnet template format. To change the full manifest
you can use files that can be read easily:
```
local deploy = import "cilium_ds.template";

deploy {
    image: "k8s1:5000/cilium/cilium-dev:latest",
    etcdEndpoint: "k8s1:9732",
    ciliumArgs: [
        "--debug=$(CILIUM_DEBUG)",
        "-t=vxlan",
        "--kvstore=etcd",
        "--kvstore-opt=etcd.config=/var/lib/etcd-config/etcd.config",
        "--disable-ipv4=$(DISABLE_IPV4)"
    ]
}
```

So to apply this one, you can do the following:
```
kubecfg show /vagrant/k8sT/manifests/cilium_ds.jsonnet | kubectl apply -f -
```

The thing is, with this tool, enable a new change in cilium like Istio config
files is straightforward and we can achieve custom manifest without all the
pain of using sed/go templates and custom tool. This 3-party tool made the work
for us.

Regards

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
Flake: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/3521/

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
@cilium/ci @cilium/janitors @cilium/contributing this is ready to review.

Regards

File: test/config/config.go [hunk 1]
```diff
@@ -38,4 +39,7 @@ func (c *CiliumTestConfigType) ParseFlags() {
 		"Specify a custom command to fetch SSH configuration (eg: 'vagrant ssh-config')")
 	flag.BoolVar(&c.ShowCommands, "cilium.showCommands", false,
 		"Output which commands are ran to stdout")
+	flag.StringVar(&c.CiliumDSManifest, "cilium.dsManifest", "",
+		"Cilium daemon set manifest to use for runing the test (only Kubernetes)")
```
Comment (ianvernon) [inline test/config/config.go]: runing --> running

File: test/helpers/cons.go [hunk 2]
```diff
@@ -177,6 +177,9 @@ const (
 
 )
 
+// CiliumDSPath the default Cilium daemon set path to use in all our test.
```
Comment (ianvernon) [inline test/helpers/cons.go]: daemon set --> DaemonSet .

"is the default ... to use in all tests"

File: test/helpers/kubectl.go [hunk 3]
```diff
@@ -714,6 +714,23 @@ func (kub *Kubectl) ApplyNetworkPolicyUsingAPI(namespace string, networkPolicy *
 	return err
 }
 
+// CiliumInstall receives a manifestName (Need to be a jsonnet format) that
```
Comment (ianvernon) [inline test/helpers/kubectl.go]: manifestName which needs to be in jsonnet format... and applied to Kubernetes.

File: test/provision/k8s_install.sh [hunk 4]
```diff
@@ -38,6 +38,9 @@ fi
 go get github.com/subfuzion/envtpl
 mv ~/go/bin/envtpl /usr/local/bin/
 
+go get github.com/ksonnet/kubecfg
```
Comment (ianvernon) [inline test/provision/k8s_install.sh]: Please update the contributing guide to include this as a dependency for those who want to run tests in their local environments.

**Review by ianvernon — CHANGES_REQUESTED** [review]

File: test/provision/k8s_install.sh (hunk 4 above)
Comment (eloycoto) [reply test/provision/k8s_install.sh]: done!

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
fixed conflicts

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
Flake: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/3641/testReport/junit/k8s-1/7/K8sValidatedPolicyTestAcrossNamespaces_Policies_Across_Namespaces/

**Comment by eloycoto** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Pull request merged**